### PR TITLE
perf: use fastpath in with_columns

### PR DESCRIPTION
**Pull request by MarcoGorelli**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
this makes things ~2x as fast, for 100_000 rows, on the `add_lags` function from scikit-lego

File: tests/utils_test.py [hunk 1]
```diff
@@ -24,6 +24,15 @@ def test_maybe_align_index_pandas() -> None:
     assert_frame_equal(nw.to_native(result), expected)
 
 
+def test_with_columns_sort() -> None:
```
Comment (MarcoGorelli) [inline tests/utils_test.py:27]: extra test for safety

File: narwhals/_pandas_like/dataframe.py [hunk 2]
```diff
@@ -167,16 +167,33 @@ def filter(
...
+        if fast_path:
+            new_names = [s.name for s in new_series]
+            df = horizontal_concat(
+                [
+                    *[
+                        self._dataframe.loc[:, s]
+                        for s in self._dataframe.columns
+                        if s not in new_names
+                    ],
+                    *[validate_dataframe_comparand(index, s) for s in new_series],
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:189]: this doesn't preserve order...one sec

**Comment by MarcoGorelli** [comment]
✅ tested this out with scikit-lego, and all the tests still pass

File: narwhals/_pandas_like/dataframe.py [hunk 3]
```diff
@@ -167,16 +167,47 @@ def filter(
 
     def with_columns(
         self,
-        *exprs: IntoPandasExpr | Iterable[IntoPandasExpr],
+        *exprs: IntoPandasExpr,
         **named_exprs: IntoPandasExpr,
     ) -> Self:
+        from narwhals._pandas_like.expr import PandasExpr
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/dataframe.py:173]: Very nit: I wonder if it's worth to import it on top level. I would expect `with_columns` to be a pretty common method

File: narwhals/_pandas_like/dataframe.py [hunk 4]
```diff
@@ -167,16 +167,47 @@ def filter(
...
         )
+
+        if fast_path:
+            new_names = {s.name: s for s in new_series}
+            to_concat = []
+            # Make sure to preserve column order
+            for s in self._dataframe.columns:
+                if s in new_names:
+                    to_concat.append(
+                        validate_dataframe_comparand(index, new_names.pop(s))
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/dataframe.py:195]: The "modifying while looping" scares the s**t out of me 🥲

File: narwhals/_pandas_like/dataframe.py [hunk 5]
```diff
@@ -167,16 +167,47 @@ def filter(
...
-        df = self._dataframe.assign(
-            **{
-                series.name: validate_dataframe_comparand(self._dataframe.index, series)
-                for series in new_series
-            }
+        # If the inputs are all Expressions which return full columns
+        # (as opposed to scalars), we can use a fast path (concat, instead of assign).
+        # We can't use the fastpath if any input is not an expression (e.g.
+        # if it's a Series) because then we might be changing its flags.
+        # See `test_memmap` for an example of where this is necessary.
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/dataframe.py:181]: Holy moly that's a great catch! 

I am getting a bit confused on the wording "e.g. if it's a Series", doesn't `validate_dataframe_comparand` check if other is a Series? (I am reading such function for the first time though)

**Review by FBruzzesi — APPROVED** [review]
Looks like an amazing feature and performance booster.

The code checks out, I left comments more for my understanding than anything else :)

File: narwhals/_pandas_like/dataframe.py (hunk 3 above)
Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:173]: yup, this works, thanks! for some reason i thought i was gonna get a circular import

in general these imports should be quite cheap, if we can reduce the circular imports that's probably a good thing

File: narwhals/_pandas_like/dataframe.py (hunk 4 above)
Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:195]: 😄 I see what you mean, but we're not modifying what we're iterating over, but another variable

we iterate over `self._dataframe.columns`, and modify `new_names`

File: narwhals/_pandas_like/dataframe.py (hunk 5 above)
Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:181]: great question! so

- expr: IntoPandasExpr. this is something that can be converted to an expression, like:
  - a string (with a column name)
  - a PandasSeries
  - a PandasExpr
- evaluate_into_expr evaluates each input `expr` into a `PandasSeries`
- `validate_dataframe_comparand` returns the underlying `pandas.Series` from each `PandasSeries`, possibly replacing the `Index`, see https://narwhals-dev.github.io/narwhals/other/pandas_index/

At this stage, we're checking whether the input `exprs` are `PandasExpr`, before they have been converted to `PandasSeries`

**Comment by MarcoGorelli** [comment]
thanks a tonne for your review!

I checked again, and this passes the scikit-lego test suite (but I'll check once more before releasing 😄 )

shipping then 🚀

**Pull request merged**